### PR TITLE
Update type inference docs

### PR DESCRIPTION
**Pull request by SupunS**
Work towards #61
⚠️ Depends on #1027

## Description

Update the documentation with the latest type-inferring rules.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by joshuahannan — APPROVED** [review]
Looks good to me!

File: docs/language/type-inference.mdx [hunk 1]
```diff
@@ -1,94 +0,0 @@
...
-```
-
-```cadence
-// Invalid: not possible to infer type based on nil literal.
-//
-let maybeSomething = nil
-
-// Instead, specify the optional type and the concrete element type, e.g. `Int`.
-//
-let maybeSomething: Int? = nil
```
Comment (turbolent) [inline docs/language/type-inference.mdx:93]: Maybe add this back to the documentation

**Review by turbolent — APPROVED** [review]
Nice!

File: docs/language/type-inference.mdx (hunk 1 above)
Comment (SupunS) [reply docs/language/type-inference.mdx:93]: This is actually valid? https://github.com/onflow/cadence/blob/32156ef08b5ab1f84c25cf23144a446717374511/docs/language/type-inference.md?plain=1#L39

Comment (turbolent) [reply docs/language/type-inference.mdx:93]: Oh, I see, the documentation was contradictory here! Thank you for cleaning it up

**Review by turbolent — APPROVED** [review]

**Pull request merged**